docs/cheat-sheets/ruby-code-injection.mdx

```diff
@@ -0,0 +1,92 @@
+---
+slug: ruby-code-injection
+---
+
+import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
+import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"
+
+# Code injection prevention for Ruby
+
+<CodeInjectionIntro />
+
+#### Check your project for these conditions:
+
+```bash
+semgrep scan --config p/ruby-command-injection
+```
+
+---
+
+## 1. Evaluating code
+
+### 1.A. Evaluating code with `eval`
+
+Evaluating code can be dangerous if dynamic content is used as input. If this input originates from outside of the program it can lead to a code injection vulnerability.
+
+Examples:
+
+```ruby
+# safe
+str = "hello"
+eval "str + ' Fred'"
+
+# vulnerable
+str = "hello"
+user_input = "system('cat /etc/passwd')" # Value supplied by user
+eval "str + #{user_input}"
+```
+
+```ruby
+class Thing
+end
+
+# safe
+Thing.module_eval(%q{def hello() "Hello there!" end})
+
+# vulnerable
+user_input = "system('cat /etc/passwd')" # Value supplied by user
+Thing.module_eval(%q{def hello() "#{user_input}" end})
+```
+
+#### References
+
+- [eval() documentation](https://www.rubydoc.info/stdlib/core/Kernel:eval)
+
+#### Mitigation
+
+- Don't use `eval()`, `class_eval()`, `module_eval()`, or `instance_eval()` if possible.
+- If you need to use `eval()`, `class_eval()`, `module_eval()`, or `instance_eval()` with non-literal values, ensure that executed content is not controllable by external sources.
+- If it's not possible, strip everything except alphanumeric characters from the input.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="ruby.lang.security.no-eval.ruby-eval" />
+
+### 1.B. Evaluating code with RubyVM::InstructionSequence
+
+The `InstructionSequence` class represents compiled instructions for the Ruby Virtual Machine. See details in [RubyVM::InstructionSequence documentation](https://ruby-doc.org/core-2.6/RubyVM/InstructionSequence.html). The `RubyVM` class itself is **not** intended for regular users. As the `RubyVM` class enables compiling code it may not potentially be secure while used with user input. Using this class with user input can result in code injection vulnerability.
```

Comment:
This passage reads awkwardly to me. Suggested change:
```suggestion
The `InstructionSequence` class represents compiled instructions for the Ruby Virtual Machine. See details in [RubyVM::InstructionSequence documentation](https://ruby-doc.org/core-2.6/RubyVM/InstructionSequence.html). The `RubyVM` class itself is **not** intended for regular users. As the `RubyVM` class enables compiling code it may insecurely interpret user input. Providing user input to this class or its methods can result in a code injection vulnerability.
```